client/src/fuzzy-search.ts

```diff
@@ -1,124 +1,57 @@
-      // for 1,000 candidates.
-      // Also, this compromise is justified in that if there that many
-      // candidates to begin with the needle is bound to match too much
-      // so the benefit of sorting by popularity is fair.
-      .slice(0, 1000);
-
-    // Now that we have a subset of only those that match, we can proceed
-    // and use `fuzzyjs` to sort by its sorting function and extract
-    // each match's ranges so we can finally result a list of sub-strings.
-    const res: CombinedMatchResult[] = candidates
-      .map((doc) => {
-        const matched = match(needle, doc.url, {
-          withScore: true,
-          withRanges: true,
-        });
-        return Object.assign(matched, doc);
-      })
-      .filter((m) => m.match && m.score && m.score > 0)
-      .sort((a, b) => {
-        // TypeScript doesn't know what we used `withScores: true`
-        // so have to do this if statement every time.
-        if (b.score && a.score) {
-          return b.score - a.score;
+  search(needle: string, { limit = 10 }): FzfResultItem<Doc>[] {
+    // Use `let` because we might come up with a new list (aka. shortlist)
+    // that makes the haystack search much simpler to send to `Fzf()`.
+    let docs = this.docs;
+    // The list of docs is possible over 10,000 entries (in 2021). If the
+    // search input is tiny, don't bother with the overhead of Fzf().
+    // Because we don't even need it when the test is so easy in that
+    // it just needs to contain a single character.
+    if (needle.length <= 3) {
+      const needleLowerCase = needle.toLowerCase();
+      // The reason this works and makes sense is because the `this.docs` is
+      // already sorted by popularity.
+      // So if someone searches for something short like `x` we just take
+      // the top 'limit' docs that have an `x` in the `.url`. This is
+      // faster than going through
+      const shortlistDocs: Doc[] = [];
+      for (const doc of this.docs) {
+        if (doc.url.toLowerCase().includes(needleLowerCase)) {
+          shortlistDocs.push(doc);
+          if (shortlistDocs.length === limit) {
+            break;
+          }
         }
-        return 0;
-      });
-    return res.slice(0, limit).map((matchResult) => {
-      return Object.assign(
-        {
-          substrings: matchResult.ranges
-            ? getSubstringsFromRanges(matchResult.url, matchResult.ranges)
-            : [],
-        },
-        matchResult
-      ) as Result;
+      }
+      // Suppose the needle was `yx`  and the `limit` as 10, then if we only found
+      // 9 (which is less than 10) docs that match exactly this, then we might
+      // be missing out, so we can't use the shortlist.
+      // For example, there might be more docs like `aaYbbbbXccc` which
+      // will be found by Fzf() but wouldn't be find in our shortlist because
+      // the two characters aren't next to each other.
+      if (shortlistDocs.length >= limit) {
```

Comment:
This was my next question after reviewing the above `shortlistDocs` implementation, regarding if the needle happened to be right underneath the `limit`. Doing a check to see if the shortlist docs is gte our upper limit then assigning `docs` the array of `Doc` objects in `shortlistDocs seems to work great. Nice work!